ios/browser/api/url/url_formatter.mm

```diff
@@ -0,0 +1,183 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/ios/browser/api/url/url_formatter.h"
+
+#include "base/bind.h"
+#include "base/callback.h"
+#include "base/notreached.h"
+#include "base/strings/sys_string_conversions.h"
+#include "build/build_config.h"
+#include "components/lookalikes/core/features.h"
+#include "components/lookalikes/core/lookalike_url_ui_util.h"
+#include "components/lookalikes/core/lookalike_url_util.h"
+#include "components/reputation/core/safety_tips_config.h"
+#include "components/url_formatter/elide_url.h"
+#include "components/url_formatter/spoof_checks/top_domains/top_domain_util.h"
+#include "components/url_formatter/url_formatter.h"
+#include "net/base/mac/url_conversions.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+
+#if !defined(__has_feature) || !__has_feature(objc_arc)
+#error "This file requires ARC support."
+#endif
+
+// MARK: - Implementation
+
+@implementation BraveURLFormatter
+
+url_formatter::SchemeDisplay IOSSchemeDisplayToCoreSchemeDisplay(BraveURLSchemeDisplay schemeDisplay) {
+  switch (schemeDisplay) {
+  case BraveURLSchemeDisplay_Show:
+    return url_formatter::SchemeDisplay::SHOW;
+  case BraveURLSchemeDisplay_OmitHttpAndHttps:
+    return url_formatter::SchemeDisplay::OMIT_HTTP_AND_HTTPS;
+  case BraveURLSchemeDisplay_OnitCryptographic:
+    return url_formatter::SchemeDisplay::OMIT_CRYPTOGRAPHIC;
+  default:
+    NOTREACHED();
+  }
+}
+
++ (bool)isPunycode:(NSString*)domain {
+  std::u16string s = base::SysNSStringToUTF16(domain);
+  return s.size() > 4 && s[0] == L'x' && s[1] == L'n' && s[2] == L'-' &&
+           s[3] == L'-';
+}
+
++ (NSString*)formatUrlForSecurityDisplay:(NSURL*)origin schemeDisplay:(BraveURLSchemeDisplay)schemeDisplay {
+  std::u16string result = url_formatter::FormatUrlForSecurityDisplay(net::GURLWithNSURL(origin), IOSSchemeDisplayToCoreSchemeDisplay(schemeDisplay));
+  return base::SysUTF16ToNSString(result);
+}
+
++ (NSString*)formatUrlForDisplayOmitSchemePathAndTrivialSubdomains:(NSURL*)url {
+  std::u16string result = url_formatter::FormatUrlForDisplayOmitSchemePathAndTrivialSubdomains(net::GURLWithNSURL(url));
+  return base::SysUTF16ToNSString(result);
+}
+@end
+
+
+// MARK: - Spoof Checker
+
+@implementation URLSpoofCheckerTopDomainEntry
+- (instancetype)initWithEntry:(url_formatter::TopDomainEntry)entry {
+  if ((self = [super init])) {
+    _domain = base::SysUTF8ToNSString(entry.domain);
+    _isTop500 = entry.is_top_500;
+    _skeletonType = static_cast<BraveSpoofCheckerSkeletonType>(entry.skeleton_type);
```

Comment:
We shouldn't static cast between 2 enums if we don't control both. Needs conversion methods